### DISCUSSION TITLE
Integration with Ardupilot compatible autopilots

### DISCUSSION THREAD

**Discussion by dfcscience** [CameronBodine/PINGMapper#107] [Ideas]
I am using a Humminbird Helix 7 G4 MEGA SI on an autonomous boogie board boat, which uses a Pixhawk for waypoint navigation. I can send the position data from the Pixhawk's GPS to the Humminbird but I would like to use the pitch, roll, and yaw data in the Pixhawk log file to improve georectification

**Comment by CameronBodine** [comment]
Thanks for your inquiry! I really dig your autonomous boogie board boat!

Correcting georectification using pitch, roll, and yaw data is a great idea. Unfortunately, that functionality is not available in PINGMapper and I don't have a lot of time to work on it at the moment. I have been giving it a think and can imagine ways to correct for each of those factors individually, but am in the dark as to how to resolve them to a final correction. If you had ideas on how the math and workflow might work, that would certainly help. I'll keep this issue open to document any ideas that come up.

**Comment by CameronBodine** [comment]
I am going to convert this to a discussion to clear out open issues. 

I may be able to start to work on this in the coming months. Do you have an example recording and the Pixhawk log file that I might be able to test any workflows that I develop?

**Reply by maxschwartziv** [reply]
Is there still interest in this?  I have a similar setup and could produce some pixhawk log files.

**Comment by maxschwartziv** [comment]
I found the hardware side of this project, looks like it would be relatively easy to get live NEMA out of the humminbird helixs.  I'd like to try this next!  @dfcscience you get this to work?

<img width="2096" height="559" alt="humminbird com port pinout" src="https://github.com/user-attachments/assets/68c65a11-0d97-4a3e-ba88-6e49d2be6cf1" />

<img width="1621" height="692" alt="echologger-ect400-pixhawk" src="https://github.com/user-attachments/assets/28a6659c-e52a-4635-8ddc-06f42b0214f2" />

<img width="595" height="711" alt="mp_rangefinder_lidarlite_testing" src="https://github.com/user-attachments/assets/382c8707-aa5b-4cb0-9890-7e2efb7f8ad6" />


https://ardupilot.org/copter/docs/common-echologger-ect400.html

https://discuss.ardupilot.org/t/arduboat-live-depth-to-gcs-map-or-reefmaster/93173/14

**Comment by maxschwartziv** [comment]
<img width="755" height="278" alt="image" src="https://github.com/user-attachments/assets/8f441ee1-c0d0-4301-a06e-5c1388a0bd57" />

So this works.

Here's the output

### Humminbird NMEA 0183 output — sentences and format

**Serial interface**

| Setting | Value |
|---|---|
| Standard | NMEA 0183 (ASCII, single talker / multiple listeners) |
| Baud rate | 4800 |
| Data / parity / stop / flow | 8 / none / 1 / none |
| Enable | Setup menu → NMEA Output = On (default Off) |
| Update rate | ~1 Hz |
| Talker ID | `IN` (integrated instrumentation) or `GP` (GPS) |

**Sentence format**

`$` + talker ID (2 chars) + sentence ID (3 chars) + `,`-separated fields + `*` + 2-hex-digit checksum (XOR of all chars between `$` and `*`) + CRLF. Empty field = no data.

**Sentences output**

| Sentence | Name | Fields | Valid data requires |
|---|---|---|---|
| DPT | Depth | Depth below transducer (m), transducer offset (m; + to waterline, − to keel) | Sonar bottom lock (fields empty without it) |
| MTW | Water temperature | Temperature, units (`C`) | Temp sensor |
| GLL | Geographic position | Lat, N/S, lon, E/W, UTC time, status (`A` valid / `V` invalid) | GPS fix |
| GGA | GPS fix data | UTC, lat, N/S, lon, E/W, fix quality (0 none, 1 GPS, 2 DGPS/SBAS), sats used, HDOP, altitude MSL + `M`, geoid separation + `M`, DGPS age, DGPS station ID | GPS fix |
| RMC | Recommended minimum GNSS data | UTC, status, lat, N/S, lon, E/W, SOG (kn), COG (°T), date `ddmmyy`, magnetic variation, E/W | GPS fix |
| VTG | Course and ground speed | COG true + `T`, COG magnetic + `M`, SOG knots + `N`, SOG km/h + `K` | GPS fix |
| ZDA | Time and date | UTC time, day, month, year, local zone hours, local zone minutes | GPS fix |
| HDG | Heading, deviation & variation | Magnetic heading, deviation, E/W, variation, E/W | Heading sensor (outputs 0.0 without one) |
| HDT | Heading, true | True heading + `T` | Heading sensor (outputs 0.0 without one) |

From the manual
<img width="561" height="287" alt="image" src="https://github.com/user-attachments/assets/37b5a899-e62a-4b40-89c1-71cf21b71a38" />
